File: clients/swift/Sources/Readiness/Features/ConfigInterfaceImplementations.swift
import PolywrapClient
import Foundation

struct ConfigInterfaceImplementationsTest: Feature {
    func runTestCase(input: Any) throws -> Void {
        guard let inputObj = input as? [AnyHashable: Any] else {
           throw InputError.expectedObject
        }

        let interfaceUri = try Input.expectUri(inputObj["interfaceUri"])
        let implementations: [String] = try Input.expectArray(inputObj["implementations"])

        print("Adding Interface Implementations to ClientConfig")

        let implementationUris = try implementations.compactMap { implementationUri -> Uri in
            guard let uri = try? Uri(implementationUri) else {
                throw UriError.parseError("Error parsing implementation uri \(implementationUri)")
            }
            return uri
        }

        let builder = BuilderConfig()
        let client = builder
            .addInterfaceImplementations(interfaceUri, implementationUris)
            .build()
        
        print("Getting Implementations")

        let result = try client.getImplementations(interfaceUri)

        if !result.isEmpty {
            print("Found \(result.count) Implementations")
            print("Success!")
        }
    }
}

File: clients/rs/src/features/config_interface_implementations.rs
use std::{error::Error};
use polywrap_client::{core::{invoker::Invoker}, client::PolywrapClient, builder::{PolywrapClientConfigBuilder, PolywrapClientConfig}};
use serde::{Deserialize};
use serde_json::Value;

use crate::input::{expect_uri, expect_array};

#[derive(Deserialize)]
struct InputObj {
  #[serde(rename = "interfaceUri")]
  interface_uri: Value,
  implementations: Value
}

pub fn run_test_case(input: &Value) -> Result<(), Box<dyn Error>> {
  let input_obj: InputObj = serde_json::from_value(input.clone())?;
  let interface_uri = expect_uri(&input_obj.interface_uri)?;
  let implementations = expect_array::<String>(&input_obj.implementations)?;
  let implementations = implementations.into_iter().map(|u| u.try_into().unwrap()).collect();

  println!("Adding Interface Implementations to ClientConfig");

  let mut config = PolywrapClientConfig::new();
  config.add_interface_implementations(interface_uri.clone(), implementations);
  
  let client: PolywrapClient = PolywrapClient::new(config.into());

  println!("Getting Implementations");

  let result = client.get_implementations(&interface_uri)?;
  let result_len = result.len();

  if result_len > 0 {
    println!("Found {result_len} Implementations");
    println!("Success!");
  }

  Ok(())
}
